diagnose.go

```diff
@@ -0,0 +1,696 @@
+	// require larger amounts of stack to deserialize. Don't increase this higher than you require.
+	MaxNestedLevels int
+
+	// MaxArrayElements specifies the max number of elements for CBOR arrays.
+	// Default is 128*1024=131072 and it can be set to [16, 2147483647]
+	MaxArrayElements int
+
+	// MaxMapPairs specifies the max number of key-value pairs for CBOR maps.
+	// Default is 128*1024=131072 and it can be set to [16, 2147483647]
+	MaxMapPairs int
+}
+
+// DiagMode returns a DiagMode with immutable options.
+func (opts DiagOptions) DiagMode() (DiagMode, error) {
+	return opts.diagMode()
+}
+
+func (opts DiagOptions) diagMode() (*diagMode, error) {
+	if err := opts.ByteStringEncoding.valid(); err != nil {
+		return nil, err
+	}
+
+	decMode, err := DecOptions{
+		MaxNestedLevels:  opts.MaxNestedLevels,
+		MaxArrayElements: opts.MaxArrayElements,
+		MaxMapPairs:      opts.MaxMapPairs,
+		// loosest decode options for diagnostic purpose.
+		UTF8: UTF8DecodeInvalid,
+	}.decMode()
+	if err != nil {
+		return nil, err
+	}
+
+	return &diagMode{&opts, decMode}, nil
+}
+
+type diagMode struct {
+	opts    *DiagOptions
+	decMode *decMode
+}
+
+// DiagOptions returns user specified options used to create this DiagMode.
+func (dm *diagMode) DiagOptions() DiagOptions {
+	return *dm.opts
+}
+
+// Diagnose returns extended diagnostic notation (EDN) of CBOR data items using the DiagMode.
+func (dm *diagMode) Diagnose(data []byte) (string, error) {
+	di, err := dm.diagnose(data)
+	if err != nil {
+		return "", err
+	}
+
+	return di.diag()
+}
+
+func (dm *diagMode) diagnose(data []byte) (*diagnose, error) {
+	de := &decoder{data: data, dm: dm.decMode}
+	off := de.off
+	err := de.valid(dm.opts.CBORSequence)
+	de.off = off
+	if err != nil {
+		return nil, err
+	}
+
+	di := &diagnose{
+		byteStringEncoding:      dm.opts.ByteStringEncoding,
+		byteStringHexWhitespace: dm.opts.ByteStringHexWhitespace,
+		byteStringText:          dm.opts.ByteStringText,
+		byteStringEmbeddedCBOR:  dm.opts.ByteStringEmbeddedCBOR,
+		cborSequence:            dm.opts.CBORSequence,
+		indicateFloatPrecision:  dm.opts.IndicateFloatPrecision,
+
+		dm: dm, d: de, w: &bytes.Buffer{},
+	}
+	return di, nil
+}
+
+var defaultDiagMode, _ = DiagOptions{}.diagMode()
+
+// Diagnose returns extended diagnostic notation (EDN) of CBOR data items
+// using the default diagnostic mode.
+//
+// Refer to https://www.rfc-editor.org/rfc/rfc8949.html#name-diagnostic-notation.
+func Diagnose(data []byte) (string, error) {
+	return defaultDiagMode.Diagnose(data)
+}
+
+type diagnose struct {
+	byteStringEncoding      ByteStringEncoding
+	byteStringHexWhitespace bool
+	byteStringText          bool
+	byteStringEmbeddedCBOR  bool
+	cborSequence            bool
+	indicateFloatPrecision  bool
+	dm                      *diagMode
+	d                       *decoder
+	w                       *bytes.Buffer
+}
+
+func (di *diagnose) diag() (string, error) {
+	if err := di.value(); err != nil {
+		return "", err
+	}
```

Comment:
See comment above that removes data validation in `diagnose()`.  So we can use this suggested `for` loop for all validation and diagnosis.

```suggestion
```